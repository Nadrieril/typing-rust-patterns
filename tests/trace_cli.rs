use assert_cmd::Command;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[test]
fn trace_cli() -> anyhow::Result<()> {
    let test_cases = &[
        "quit",
        "help",
        "options",
        "set",
        "set mistyped",
        "set mistyped val",
        "set ref_binding_on_inherited true",
        "set allow_ref_pat_on_ref_mut false",
        "set downgrade_shared_inside_shared true",
        "set ergo2024\noptions",
        "&x: &mut T",
        "set allow_ref_pat_on_ref_mut false\n&x: &mut T",
        "rules",
    ];
    for &test_case in test_cases {
        let success = Command::cargo_bin("typing-rust-patterns")?
            .write_stdin(test_case)
            .assert()
            .success();
        let out = String::from_utf8(success.get_output().stdout.clone())?;

        // Identify each snapshot file by the hash of the test case.
        let req_hash = {
            let mut hasher = DefaultHasher::new();
            test_case.hash(&mut hasher);
            hasher.finish().to_string()
        };
        insta::with_settings!({
            snapshot_suffix => req_hash,
            info => &test_case,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(out);
        });
    }
    Ok(())
}
