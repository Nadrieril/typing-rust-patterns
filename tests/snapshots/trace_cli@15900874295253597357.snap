---
source: tests/trace_cli.rs
info: save;set fallback_to_outer EatOuter;set inherited_ref_on_ref EatInner
---
Current ruleset was saved
The two rulesets are described by the following sets of rules, with differences highlighted.
The old ruleset is on the left, and the new one on the right.


allow_ref_pat_on_ref_mut: true
downgrade_mut_inside_shared: false
eat_inherited_ref_alone: true
eat_mut_inside_shared: true
fallback_to_outer: "No" -> "EatOuter"
inherited_ref_on_ref: "EatOuter"
match_constructor_through_ref: true
mut_binding_on_inherited: "Keep"
ref_binding_on_inherited: "Error"
simplify_deref_mut: true
The two rulesets are described by the following sets of rules, with differences highlighted.
The old ruleset is on the left, and the new one on the right.

 real, ro ⊢ p: &T                                                                 | inh, ro ⊢ p: &T
 ---------------- "Deref"                                                         | ---------------- "Deref(EatInner)"
 inh, m ⊢ &p: &&T                                                                 | inh, m ⊢ &p: &&T

 real, ro ⊢ p: &mut T                                                             | inh, ro ⊢ p: &T
 -------------------- "Deref"                                                     | -------------------- "DerefMutWithShared(EatInner)"
 inh, m ⊢ &p: &&mut T                                                             | inh, m ⊢ &p: &&mut T

 real, m ⊢ p: &mut T                                                              | inh, m ⊢ p: &mut T
 ---------------------------- "Deref"                                             | ---------------------------- "Deref(EatInner)"
 inh, m ⊢ &mut p: &mut &mut T                                                     | inh, m ⊢ &mut p: &mut &mut T

                                                                                  | inh, m ⊢ p: &T
                                                                                  | ------------------------ "Deref(EatInner)"
                                                                                  | inh, m ⊢ &mut p: &&mut T

 real, ro ⊢ p: &T                                                                 | inh, ro ⊢ p: &mut T
 -------------------- "DerefMutWithShared"                                        | -------------------- "Deref(EatInner)"
 inh, m ⊢ &p: &mut &T                                                             | inh, m ⊢ &p: &mut &T

 real, ro ⊢ p: &mut T                                                             | inh, ro ⊢ p: &mut T
 ------------------------ "DerefMutWithShared"                                    | ------------------------ "DerefMutWithShared(EatInner)"
 inh, m ⊢ &p: &mut &mut T                                                         | inh, m ⊢ &p: &mut &mut T


allow_ref_pat_on_ref_mut: true
downgrade_mut_inside_shared: false
eat_inherited_ref_alone: true
eat_mut_inside_shared: true
fallback_to_outer: "EatOuter"
inherited_ref_on_ref: "EatOuter" -> "EatInner"
match_constructor_through_ref: true
mut_binding_on_inherited: "Keep"
ref_binding_on_inherited: "Error"
simplify_deref_mut: true
